File: miso/code/win32_handmade.m
-|                  |
 | Handmade Hero    |
 |                  |-

#import "Windows" as Win32

Win32OffscreenBuffer: {
    info            Win32.BITMAPINFO
    memory          ^void
    width           int
    height          int
    pitch           int
    bytesPerPixel   int
}

var 
    gRunning = true,
    gGlobalBackBuffer: Win32OffscreenBuffer


Win32GetWindowDimension: (window HWND) -> (width int, height int)
{
    var clientRect: RECT
    Win32.GetClientRect(window, ^clientRect)
    return (clientRect.right - clientRect.left, clientRect.bottom - clientRect.top)
}

RenderWeirdGradient: (buffer Win32OffscreenBuffer, blueOffset int, greenOffset int)
{
    let row = buffer.memory as ^u8
    loop y in (0, buffer.height)
    {
        unsafe
        {
            var pixel: ^u8 = row
            loop x in (0, buffer.width)
            {
                let blue = (x + blueOffset) as u8
                let green = (y + greenOffset) as u8

                pixel^= (green as u32) << 8 | (blue as u32)
                pixel += 1
            }

            row += buffer.pitch
        }
    }
}

Win32ResizeDIBSection: (buffer ^Win32OffscreenBuffer mutable, width int, height int)
{
    if buffer^.memory Win32.VirtualFree(gBitmapMemory, 0, Win32.MEM_RELEASE)

    buffer^.width = width;
    buffer^.height = height;

    buffer^.info.bmiheader = {
        biSize:         sizeof(buffer^.info.bmiHeader)
        biWidth:        buffer^.width
        biHeight:       -buffer^.height
        biPlanes:       1
        biBitCount:     32
        biCompression:  Win32.BI_RGB
    }

    let bitmapMemorySize = (buffer^.width * buffer^.height) * buffer^.bytesPerPixel
    buffer^.memory = Win32.VirtualAlloc(null, bitmapMemorySize, Win32.MEM_COMMIT, Win32.PAGE_READWRITE)
}

Win32UpdateWindow: (dc Win32.HDC, windowWidth int, windowHeight int, buffer Win32OffscreenBuffer, x int, y int, width int, height int)
{
    Win32.StretchDIBits(
        dc,
        0, 0, windowWidth, windowHeight
        x, y, buffer.width, buffer.height
        buffer.memory,
        ^buffer.info,
        Win32.DIB_RGB_COLORS,
        Win32.SRCCOPY)
}

main: ()
{
    Win32ResizeDIBSection(^gGlobalBackBuffer, 1280, 720)

    let windowClass: Win32.WNDCLASS = {
        style:          Win32.CS_HREDRAW | Win32.CS_VREDRAW
        lpfnWndProc:    (window Win32.HWND, message Win32.UINT, wParam Win32.WPARAM, lParam Win32.LPARAM) -> LRESULT
                        {
                            var result: Win32.LRESULT = 0

                            select message
                            {
                                on Win32.WM_SIZE
                                {
                                }

                                on Win32.WM_CLOSE
                                {
                                    gRunning = false
                                }

                                on Win32.WM_ACTIVATEAPP
                                {

                                }

                                on Win32.WM_DESTROY
                                {
                                    gRunning = false
                                }

                                on Win32.WM_PAINT
                                {
                                    var paint: Win32.PAINTSTRUCT
                                    let dc = Win32.BeginPaint(window, ^paint)

                                    let x = paint.rcPaint.left as int
                                    let y = paint.rcPaint.top as int
                                    let width = paint.rcPaint.right - x as int
                                    let height = paint.rcPaint.bottom - y as int

                                    let windowWidth, windowHeight = Win32GetWindowDimension(window)
                                    Win32UpdateWindow(dc, windowWidth, windowHeight, gGlobalBackBuffer, x, y, width, height)

                                    Win32.EndPaint(window, ^paint)
                                }

                                default
                                {
                                    result = DefWindowProc(window, message, wParam, lParam)
                                }
                            }
                        }
        hInstance:      Win32.GetModuleHandle(null)
        lpszClassName:  "HandmadeHeroWindowClass"
    }

    if Win32.RegisterClass(^windowClass)
    {
        let window = Win32.CreateWindowEx(
            lpszClassName:      windowClassName.lpszClassName,
            lpszWindowName:     "Handmade Hero",
            dwStyle:            Win32.WS_OVERLAPPEDWINDOW | Win32.WS_VISIBLE,
            x:                  CW_USEDEFAULT,
            y:                  CW_USEDEFAULT,
            nWidth:             CW_USEDEFAULT,
            nHeight:            CW_USEDEFAULT,
            hInstance:          windowClassName.hInstance
        )

        if (window)
        {
            var xOffset, yOffset: int
            gRunning = true
            
            while(gRunning) {
                var message: Win32.MSG
                while(Win32.PeekMessage(^message, 0, 0, 0, Win32.PM_REMOVE))
                {
                    gRunning = false if (message.message == Win32.WM_QUIT)
                    Win32.TranslateMessage(^message)
                    Win32.DispatchMessage(^message)
                }

                RenderWeirdGradient(gGlobalBackBuffer, xOffset, yOffset)

                let deviceContext = Win32.GetDC(window)
                let windowWidth, windowHeight = Win32GetWindowDimension(window)
                Win32UpdateWindow(deviceContext, windowWidth, windowHeight, gGlobalBackBuffer, 0, 0, windowWidth, windowHeight)
                Win32.ReleaseDC(window, deviceContext)

                xOffset += 1
                yOffset += 2
            }
        }
    }
    else
    {

    }

    return 0
}

File: c/code/win32_handmade.cc
//
// Handmade Hero
//

#include <Windows.h>
#include <stdint.h>
#include <Xinput.h>
#include <dsound.h>

#define internal static
#define local_persist static
#define global_variable static

//----------------------------------------------------------------------------------------------------------------------
// Basic types

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;
typedef uint64_t u64;

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;
typedef int64_t s64;

typedef int32_t b32;
typedef int64_t b64;

//----------------------------------------------------------------------------------------------------------------------
// Win32OffscreenBuffer

struct Win32OffscreenBuffer
{
    BITMAPINFO info;
    void* memory;
    int width;
    int height;
    int pitch;
};

//----------------------------------------------------------------------------------------------------------------------
// XInput API


#define X_INPUT_GET_STATE(name) DWORD WINAPI name(DWORD dwUserIndex, XINPUT_STATE* pState)
typedef X_INPUT_GET_STATE(XInputGetStateFunc);
X_INPUT_GET_STATE(XInputGetStateStub)
{
    return ERROR_DEVICE_NOT_CONNECTED;
}
global_variable XInputGetStateFunc* XInputGetState_ = XInputGetStateStub;

#define X_INPUT_SET_STATE(name) DWORD WINAPI name(DWORD dwUserIndex, XINPUT_VIBRATION* pVibration)
typedef X_INPUT_SET_STATE(XInputSetStateFunc);
X_INPUT_SET_STATE(XInputSetStateStub)
{
    return ERROR_DEVICE_NOT_CONNECTED;
}
global_variable XInputSetStateFunc* XInputSetState_ = XInputSetStateStub;

#define XInputGetState XInputGetState_
#define XInputSetState XInputSetState_

internal void Win32LoadXInput()
{
    HMODULE xInputLibrary = LoadLibraryA("xinput1_4.dll");
    if (!xInputLibrary)
    {
        xInputLibrary = LoadLibraryA("xinput1_3.dll");
    }

    if (xInputLibrary)
    {
        XInputGetState = (XInputGetStateFunc*)GetProcAddress(xInputLibrary, "XInputGetState");
        if (!XInputGetState) XInputGetState = XInputGetStateStub;

        XInputSetState = (XInputSetStateFunc*)GetProcAddress(xInputLibrary, "XInputSetState");
        if (!XInputSetState) XInputSetState = XInputSetStateStub;
    }
}

//----------------------------------------------------------------------------------------------------------------------
// Sound management

#define DIRECT_SOUND_CREATE(name) HRESULT WINAPI name(LPCGUID pcGuidDevice, LPDIRECTSOUND* ppDS, LPUNKNOWN pUnkOuter)
typedef DIRECT_SOUND_CREATE(DirectSoundCreateFunc);

internal void Win32InitDSound(HWND window, u32 samplesPerSec, u32 bufferSize)
{
    // Load the library
    HMODULE dSoundLibrary = LoadLibraryA("dsound.dll");
    if (dSoundLibrary)
    {
        // Get a DirectSound object
        DirectSoundCreateFunc* DirectSoundCreate = (DirectSoundCreateFunc*)GetProcAddress(dSoundLibrary, 
                                                                                          "DirectSoundCreate");
        LPDIRECTSOUND directSound;
        if (DirectSoundCreate && SUCCEEDED(DirectSoundCreate(0, &directSound, 0)))
        {
            WAVEFORMATEX waveFormat = {};
            waveFormat.wFormatTag = WAVE_FORMAT_PCM;
            waveFormat.nChannels = 2;
            waveFormat.wBitsPerSample = 16;
            waveFormat.nSamplesPerSec = samplesPerSec;
            waveFormat.nBlockAlign = waveFormat.nChannels * waveFormat.wBitsPerSample / 8;
            waveFormat.nAvgBytesPerSec = waveFormat.nSamplesPerSec * waveFormat.nBlockAlign;
            waveFormat.cbSize = 0;

            if (SUCCEEDED(directSound->SetCooperativeLevel(window, DSSCL_PRIORITY)))
            {
                // Create a primary buffer
                DSBUFFERDESC bufferDescription = { };
                bufferDescription.dwSize = sizeof(bufferDescription);
                bufferDescription.dwFlags = DSBCAPS_PRIMARYBUFFER;

                LPDIRECTSOUNDBUFFER primaryBuffer;
                if (SUCCEEDED(directSound->CreateSoundBuffer(&bufferDescription, &primaryBuffer, 0)))
                {
                    if (SUCCEEDED(primaryBuffer->SetFormat(&waveFormat)))
                    {
                        DSBUFFERDESC bufferDescription = {};
                        bufferDescription.dwSize = sizeof(bufferDescription);
                        bufferDescription.dwFlags = 0;
                        bufferDescription.dwBufferBytes = bufferSize;
                        bufferDescription.lpwfxFormat = &waveFormat;

                        LPDIRECTSOUNDBUFFER secondaryBuffer;
                        if (SUCCEEDED(directSound->CreateSoundBuffer(&bufferDescription, &secondaryBuffer, 0)))
                        {
                            OutputDebugStringA("Created secondary sound buffer\n");
                        }
                    }
                    else
                    {
                        OutputDebugStringA("ERROR: Could not set format on primary buffer\n");
                    }
                }
                else
                {
                    OutputDebugStringA("ERROR: Could not create primary buffer");
                }
            }
            else
            {
                OutputDebugStringA("ERROR: Could not set cooperative level");
            }
        }
    }
}

//----------------------------------------------------------------------------------------------------------------------
// Global state

global_variable bool gRunning = true;
global_variable Win32OffscreenBuffer gGlobalBackBuffer;

//----------------------------------------------------------------------------------------------------------------------
// Win32GetWindowDimension

struct Win32WindowDimension
{
    int width;
    int height;
};

Win32WindowDimension Win32GetWindowDimension(HWND window)
{
    RECT clientRect;
    GetClientRect(window, &clientRect);
    return{
        clientRect.right - clientRect.left,
        clientRect.bottom - clientRect.top
    };
}

//----------------------------------------------------------------------------------------------------------------------
// Rendering

internal void RenderWeirdGradient(Win32OffscreenBuffer* buffer, int blueOffset, int greenOffset)
{
    u8* row = (u8 *)buffer->memory;
    for (int y = 0; y < buffer->height; ++y)
    {
        u32* pixel = (u32 *)row;
        for (int x = 0; x < buffer->width; ++x)
        {
            // Pixel in memory: 00 RR GG BB
            u8 blue = (x + blueOffset);
            u8 green = (y + greenOffset);

            *pixel++ = (((u32)green << 8) | blue);
        }

        row += buffer->pitch;
    }
}

//----------------------------------------------------------------------------------------------------------------------
// Backbuffer creation

internal void Win32ResizeDIBSection(Win32OffscreenBuffer* buffer, int width, int height)
{
    if (buffer->memory)
    {
        VirtualFree(buffer->memory, 0, MEM_RELEASE);
    }

    buffer->width = width;
    buffer->height = height;
    int bytesPerPixel = 4;

    // Create the new DIBSection
    buffer->info.bmiHeader.biSize = sizeof(buffer->info.bmiHeader);
    buffer->info.bmiHeader.biWidth = buffer->width;
    buffer->info.bmiHeader.biHeight = -buffer->height;
    buffer->info.bmiHeader.biPlanes = 1;
    buffer->info.bmiHeader.biBitCount = 32;
    buffer->info.bmiHeader.biCompression = BI_RGB;

    int bitmapMemorySize = (buffer->width * buffer->height) * bytesPerPixel;
    buffer->memory = VirtualAlloc(0, bitmapMemorySize, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);

    buffer->pitch = width * bytesPerPixel;

}

//----------------------------------------------------------------------------------------------------------------------
// Win32 rendering

internal void Win32DisplayBufferInWindow(Win32OffscreenBuffer* buffer, HDC dc, int windowWidth, int windowHeight)
{
    StretchDIBits(
        dc,
        0, 0, windowWidth, windowHeight,
        0, 0, buffer->width, buffer->height,
        buffer->memory,
        &buffer->info,
        DIB_RGB_COLORS,
        SRCCOPY);
}

//----------------------------------------------------------------------------------------------------------------------
// Win32 messages handler

LRESULT CALLBACK Win32MainWindowCallback(HWND window, 
                                         UINT message,
                                         WPARAM wParam,
                                         LPARAM lParam)
{
    LRESULT result = 0;

    switch(message)
    {
    case WM_CLOSE:
        {
            gRunning = false;
        }
        break;

    case WM_ACTIVATEAPP:
        {
            OutputDebugStringA("WM_ACTIVATEAPP\n");
        }
        break;

    case WM_DESTROY:
        {
            gRunning = false;
        }
        break;

    case WM_SYSKEYDOWN:
    case WM_SYSKEYUP:
    case WM_KEYDOWN:
    case WM_KEYUP:
        {
            u32 vkCode = (u32)wParam;
            bool wasDown = ((lParam & (1 << 30)) != 0);
            bool isDown = ((lParam & (1 << 31)) == 0);

            if (wasDown != isDown)
            {
                if (vkCode == 'W')
                {
                }
                else if (vkCode == 'A')
                {
                }
                else if (vkCode == 'S')
                {
                }
                else if (vkCode == 'D')
                {
                }
                else if (vkCode == 'Q')
                {
                }
                else if (vkCode == 'E')
                {
                }
                else if (vkCode == VK_UP)
                {
                }
                else if (vkCode == VK_LEFT)
                {
                }
                else if (vkCode == VK_DOWN)
                {
                }
                else if (vkCode == VK_RIGHT)
                {
                }
                else if (vkCode == VK_ESCAPE)
                {
                }
                else if (vkCode == VK_SPACE)
                {
                }
            }

            b64 altKeyWasDown = (lParam & (1ull << 29));
            if ((vkCode == VK_F4) && altKeyWasDown)
            {
                gRunning = false;
            }
        }
        break;

    case WM_PAINT:
        {
            PAINTSTRUCT paint;
            HDC dc = BeginPaint(window, &paint);
            Win32WindowDimension dimension = Win32GetWindowDimension(window);
            Win32DisplayBufferInWindow(&gGlobalBackBuffer, dc, dimension.width, dimension.height);
            EndPaint(window, &paint);
        }
        break;

    default:
        {
            result = DefWindowProc(window, message, wParam, lParam);
        }
        break;
    }

    return result;
}

//----------------------------------------------------------------------------------------------------------------------
// Main loop

int CALLBACK WinMain(HINSTANCE inst, 
                     HINSTANCE prevInst, 
                     LPSTR cmdLine, 
                     int cmdShow)
{
    Win32LoadXInput();
 
    WNDCLASSA windowClass = {};

    Win32ResizeDIBSection(&gGlobalBackBuffer, 1280, 720);

    windowClass.style = CS_HREDRAW | CS_VREDRAW | CS_OWNDC;
    windowClass.lpfnWndProc = &Win32MainWindowCallback;
    windowClass.hInstance = inst;
    windowClass.lpszClassName = "HandmadeHeroWindowClass";

    if (RegisterClass(&windowClass))
    {
        HWND window = CreateWindowEx(
            0,
            windowClass.lpszClassName,
            "Handmade Hero",
            WS_OVERLAPPEDWINDOW | WS_VISIBLE,
            CW_USEDEFAULT,
            CW_USEDEFAULT,
            CW_USEDEFAULT,
            CW_USEDEFAULT,
            0,
            0,
            inst,
            0);
        if (window)
        {
            HDC deviceContext = GetDC(window);

            int xOffset = 0;
            int yOffset = 0;

            Win32InitDSound(window, 48000, 48000*sizeof(s16)*2);

            gRunning = true;
            while(gRunning)
            {
                MSG message;
                while(PeekMessage(&message, 0, 0, 0, PM_REMOVE))
                {
                    if (message.message == WM_QUIT)
                    {
                        gRunning = false;
                    }

                    TranslateMessage(&message);
                    DispatchMessage(&message);
                }

                for (DWORD controllerIndex = 0; controllerIndex <= XUSER_MAX_COUNT; ++controllerIndex)
                {
                    XINPUT_STATE controllerState;
                    if (XInputGetState(controllerIndex, &controllerState) == ERROR_SUCCESS)
                    {
                        // This controller is plugged in.
                        XINPUT_GAMEPAD* pad = &controllerState.Gamepad;

                        bool up = (pad->wButtons & XINPUT_GAMEPAD_DPAD_UP) != 0;
                        bool down = (pad->wButtons & XINPUT_GAMEPAD_DPAD_DOWN) != 0;
                        bool left = (pad->wButtons & XINPUT_GAMEPAD_DPAD_LEFT) != 0;
                        bool right = (pad->wButtons & XINPUT_GAMEPAD_DPAD_RIGHT) != 0;
                        bool start = (pad->wButtons & XINPUT_GAMEPAD_START) != 0;
                        bool back = (pad->wButtons & XINPUT_GAMEPAD_BACK) != 0;
                        bool leftShoulder = (pad->wButtons & XINPUT_GAMEPAD_LEFT_SHOULDER) != 0;
                        bool rightShoulder = (pad->wButtons & XINPUT_GAMEPAD_RIGHT_SHOULDER) != 0;
                        bool aButton = (pad->wButtons & XINPUT_GAMEPAD_A) != 0;
                        bool bButton = (pad->wButtons & XINPUT_GAMEPAD_B) != 0;
                        bool xButton = (pad->wButtons & XINPUT_GAMEPAD_X) != 0;
                        bool yButton = (pad->wButtons & XINPUT_GAMEPAD_Y) != 0;

                        s16 stickX = pad->sThumbLX;
                        s16 stickY = pad->sThumbLY;

                        xOffset += stickX >> 12;
                        yOffset += stickY >> 12;
                    }
                    else
                    {
                        // This controller is not available
                    }
                }

                RenderWeirdGradient(&gGlobalBackBuffer, xOffset, yOffset);

                Win32WindowDimension dimension = Win32GetWindowDimension(window);
                Win32DisplayBufferInWindow(&gGlobalBackBuffer, deviceContext, dimension.width, dimension.height);

                ++xOffset;
            }
        }
    }
    else
    {

    }

    return 0;
}

